src/utils/avatar.js

```diff
@@ -314,14 +314,32 @@ export class UploadedAvatarURL extends AvatarURL {
    * Expects a relative URL plus the realm for a local upload;
    * otherwise, an absolute URL of the avatar on the S3 backend.
    */
+  // We should avoid doing unnecessary `new URL` calls here. They are
+  // very expensive, and their use in these pseudo-constructors (which
+  // process data at the edge, just as it's received from the server)
+  // used to slow down `api.registerForEvents` quite a lot.
   static validateAndConstructInstance(args: {|
     realm: URL,
     absoluteOrRelativeUrl: string,
   |}): UploadedAvatarURL {
     const { realm, absoluteOrRelativeUrl } = args;
     // If `absoluteOrRelativeUrl` is absolute, the second argument
     // is ignored.
-    return new UploadedAvatarURL(new URL(absoluteOrRelativeUrl, realm));
+    return new UploadedAvatarURL(
+      // If it's relative, it's probably easiest (for now, at least)
+      // just to do the `new URL` call. If we want to defer that, then
+      // we'll have to reimplement `serialize` and `deserialize` to
+      // not depend on `this._standardUrl` being present, and that
+      // might get messy.
+      //
+      // If it's absolute, we can defer `new URL` to the first
+      // `.get()` call: we should be able to trust that the server
+      // gives us properly formatted URLs; if it didn't, it seems like
+      // the kind of bug that would be fixed quickly.
+      absoluteOrRelativeUrl.startsWith('/')
+        ? new URL(absoluteOrRelativeUrl, realm)
+        : absoluteOrRelativeUrl,
```

Comment:
A couple of thoughts on this one:
* This logic is completely correct that if the URL string starts with a `/`, it's definitely a relative-URL string (provided it's any kind of valid URL string, and I guess has no fragment). With just a bit more logic, I think we can actually be more specific: we can make it definitely a ["path-absolute-URL string"](https://url.spec.whatwg.org/#path-absolute-url-string), and then we can say `${realm.origin}${absoluteOrRelativeURL}` and skip the `new URL` here too.
* If the string doesn't start with a slash, it isn't necessarily absolute; it could be one of a few other flavors of relative-URL strings. I think *probably* a well-behaved server will never want to give us one of those... but I'm not sure of that. So in the spirit of the crunchy shell, I'd like to validate it just a little more to be sure it's absolute, and if it's neither absolute nor the flavor of relative that we expect, then give an error.

How about I try developing a commit for this uploaded-avatar case? I have the [URL Standard](https://url.spec.whatwg.org/#url-parsing) open, and I think I can make something short and simple that reliably does the small part of it we need here.

Reply:
> How about I try developing a commit for this uploaded-avatar case? I have the [URL Standard](https://url.spec.whatwg.org/#url-parsing) open, and I think I can make something short and simple that reliably does the small part of it we need here.

Cool, sounds good, thanks! 🙂